pysingfel/experiment/holo.py

```diff
@@ -47,51 +26,87 @@ def generate_new_sample_state(self):
-                part_orientations = psg.get_random_quat(n_particles)
-                part_positions = part_positions[part_states == self.particles[i]]
-        else:
-            # particle cluster
-            particle_distribution = np.random.multinomial(
-                self.n_part_per_shot, self.part_ratios)
-            part_dict = {self.particles[i]: n_particles for i, n_particles in enumerate(particle_distribution)}
-            part_states, part_positions = distribute_particles(ref_dict, self.beam.get_focus()[0]/2, jet_radius=1e-4, gamma=1.)
-            part_states = np.array(part_states)
-            for i in range(self.n_particle_kinds):
-                n_particles = particle_distribution[i]
-                part_orientations = psg.get_random_quat(n_particles)
-                part_positions = part_positions[part_states == self.particles[i]]
-            # reference cluster
-            reference_distribution = np.random.multinomial(
-                self.n_ref_per_shot, self.ref_ratios)
-            ref_dict = {self.reference[i]: n_reference for i, n_reference in enumerate(reference_distribution)}
-            ref_states, reference_positions = distribute_particles(ref_dict, self.beam.get_focus()[0]/2, jet_radius=1e-4, gamma=1.)
-            ref_positions[:,1] = part_positions[:ref_positions.shape[0],1]
-            ref_states = np.array(reference_states)
-            for i in range(self.n_reference_kinds):
-                n_reference = reference_distribution[i]
-                ref_orientations = psg.get_random_quat(n_reference)
-                ref_positions = ref_positions[ref_states == self.reference[i]]
-        positions = np.concatenate((ref_positions,part_positions), axis=0)
-        orientations = np.concatenate((ref_orientations,part_orientations), axis=0)
+        part_orientations = self.get_next_part_orientation()
+        part_positions = self.get_next_part_position()
+        print("part_positions =", part_positions)
+        ref_orientation = self.get_ref_orientation()
+        ref_position = self.get_ref_position()
+        print("ref_position =", ref_position)
+        positions = np.concatenate((ref_position,part_positions), axis=0)
+        orientations = np.concatenate((ref_orientation,part_orientations), axis=0)
         particle_groups.append((positions, orientations))
+        print("particle_groups =", particle_groups)
         return particle_groups
+
+    def get_ref_orientation(self):
+        """
+        Return the orientation.
+        """
+        if self._ref_orientation is None:
+            return psg.get_random_quat(1)
+
+        ref_orientation = self._ref_orientation
+
+        return ref_orientation
+
+    def set_ref_orientation(self, ref_orientation):
+        self._ref_orientation = ref_orientation
+
+    def get_ref_position(self):
+        """
+        Return the position.
+        """
+        if self._ref_position is None:
+            return np.array([[0., -1e-6, 0.]])
```

Comment:
is there a reason to set it to -1e-6? [0,0,0] seems more logical

Reply:
The default position of the sample particle is at [0,0,0]. I thought we want to avoid placing the reference particle at the same position.

Reply:
If _ref_position is None, I think making up some position isn't correct behavior. The method is a "get" method, but you are setting a value and returning here. Why not just return None?

I don't have a deep understanding of your code and I'm just commenting on what I can see. So take with grain of salt.

Reply:
Can this return None instead? I think that would be the correct behavior for this get function.
You are not setting self._ref_position a value before returning this value

Reply:
these kind of changes can break code, so it would be good to include in pytest